common/Storyboarding/OsbSpriteWriter.cs

```diff
@@ -0,0 +1,306 @@
+            OsbLayer = layer;
+        }
+
+        public void WriteOsb()
+        {
+            if (OsbSprite.CommandCount == 0)
+                return;
+
+            if (OsbSprite.CommandCount > OsbSprite.MaxCommandCount && IsFragmentable())
+            {
+                HashSet<int> fragmentationTimes = GetFragmentationTimes();
+
+                List<ICommand> commands = OsbSprite.Commands.ToList();
+
+                while (commands.Count > 0)
+                {
+                    var segment = GetNextSegment(fragmentationTimes, commands);
+                    var sprite = CreateSprite(segment);                       
+                    WriteOsbSprite(sprite);
+                }
+            }
+            else
+            {
+                WriteOsbSprite(OsbSprite);
+            }
+        }
+
+        protected virtual OsbSprite CreateSprite(List<ICommand> segment)
+        {
+            var sprite = new OsbSprite()
+            {
+                TexturePath = OsbSprite.TexturePath,
+                InitialPosition = OsbSprite.InitialPosition,
+                Origin = OsbSprite.Origin,
+            };
+
+            foreach (var command in segment)
+                sprite.AddCommand(command);
+
+            return sprite;
+        }
+
+        private void WriteOsbSprite(OsbSprite sprite)
+        {
+            WriteHeader(sprite);
+            foreach (var command in sprite.Commands)
+                command.WriteOsb(TextWriter, ExportSettings, 1);
+        }
+
+        protected virtual void WriteHeader(OsbSprite sprite)
+        {
+            TextWriter.Write($"Sprite,{OsbLayer},{sprite.Origin},\"{sprite.TexturePath.Trim()}\"");
+            if (!moveTimeline.HasCommands && !moveXTimeline.HasCommands)
+                TextWriter.Write($",{sprite.InitialPosition.X.ToString(ExportSettings.NumberFormat)}");
+            else TextWriter.Write($",0");
+            if (!moveTimeline.HasCommands && !moveYTimeline.HasCommands)
+                TextWriter.WriteLine($",{sprite.InitialPosition.Y.ToString(ExportSettings.NumberFormat)}");
+            else TextWriter.WriteLine($",0");
+        }
+
+        protected virtual bool IsFragmentable()
+        {
+            if (OsbSprite.CommandCount < OsbSprite.MaxCommandCount)
+                return false;
+
+            return !(moveTimeline.HasOverlap ||
+                     moveXTimeline.HasOverlap ||
+                     moveYTimeline.HasOverlap ||
+                     rotateTimeline.HasOverlap ||
+                     scaleTimeline.HasOverlap ||
+                     scaleVecTimeline.HasOverlap ||
+                     fadeTimeline.HasOverlap ||
+                     colorTimeline.HasOverlap);
+        }
+
+        protected virtual HashSet<int> GetFragmentationTimes()
+        {
+            HashSet<int> fragmentationTimes = new HashSet<int>();
+            var nonFragmentableCommands = OsbSprite.Commands.Where(c => !c.IsFragmentable()).ToList();

```

Comment:
that .ToList() seems unnecessary?

Reply:
yes, changed it and adjusted the foreach loop accordingly